src/it/scala/com/wavesplatform/it/ValidChainGenerationSpec.scala

```diff
@@ -0,0 +1,44 @@
+package com.wavesplatform.it
+
+import org.scalatest._
+import org.scalatest.concurrent.{Eventually, IntegrationPatience, ScalaFutures}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.util.Random
+
+class ValidChainGenerationSpec(allNodes: Seq[Node]) extends FreeSpec with ScalaFutures with IntegrationPatience
+  with Matchers with Eventually {
+
+  "Generate 30 blocks and synchronise" in {
+    val addressToNode = allNodes.map(n => n.address -> n).toMap
+    var balances = Await.result(Future.traverse(allNodes)(n => n.balance(n.address))
+      .map(_.map(b => b.address -> b.balance).toMap), 30.seconds)
+
+    def makeTransfer(): Future[Node.Transaction] = {
+      val shuffledAddresses = Random.shuffle(addressToNode.keys)
+      val sourceAddress = shuffledAddresses.head
+      val sourceNode = addressToNode(sourceAddress)
+      val targetAddress = Random.shuffle(shuffledAddresses.tail).head
+      val fee = 100000
+      val amount = (Random.nextDouble() * 1e-3 * (balances(sourceAddress) - fee)).toLong
+
+      balances += sourceAddress -> (balances(sourceAddress) - amount)
```

Comment:
As the test is written now, it takes too much attention to understand, what is it actually doing.
The busniess logic of transfer is interleaved with the building transfer itself and some state of some `var`.

I understand this might be some "special" complex test, but if we want this to be a sort of reference implementation, we need to simplify it -- or at least try less mutable approach.


Regarding the random, itself, I'd give another shot to generators since they do compose.